### ISSUE TITLE
How is the weighted matrix trained?

### ISSUE THREAD

**Issue by KHN190** [karpathy/char-rnn#100]
Hi Karpathy,

I read in the code of GRU the weight of gates seems to be randomly generated every time you push through the inputs:

    -- forward the update and reset gates
    local update_gate = nn.Sigmoid()(new_input_sum(input_size_L, x, prev_h))
    local reset_gate = nn.Sigmoid()(new_input_sum(input_size_L, x, prev_h))
    -- compute candidate hidden state
    local gated_hidden = nn.CMulTable()({reset_gate, prev_h})
    local p2 = nn.Linear(rnn_size, rnn_size)(gated_hidden)
    local p1 = nn.Linear(input_size_L, rnn_size)(x)
    local hidden_candidate = nn.Tanh()(nn.CAddTable()({p1,p2}))

The weights and bias in `nn.Linear()` don't seem to be held somewhere and decay. Same mechanism can be found in LSTM ([wojzaremba/lstm](https://github.com/wojzaremba/lstm/blob/master/main.lua#L65)) model:

    local function lstm(x, prev_c, prev_h)
      local i2h = nn.Linear(params.rnn_size, 4*params.rnn_size)(x)
      local h2h = nn.Linear(params.rnn_size, 4*params.rnn_size)(prev_h)
      -- some stuff here
      -- ..
      return next_c, next_h
    end

, where seems each time you forward `x, prev_c, prev_h` through LSTM, the weights are newly created. So how could the weights are trained to optimize? 

That is saying, if I'm going to create a global reset gate described in GF-RNN, it will be right to do the following (right: means the weight can be trained):

    -- create a global reset gate
    function gr_gate(h_t, h_s)

	  -- assume it's a matrix
	  local t_size = h_t:size(2)
	  local s_size = h_s:size(2)
	  -- weighted hidden state
	  local w_t = nn.Linear( t_size, t_size )(h_t)
	  local w_s = nn.Linear( s_size, t_size )(h_s)

	  return nn.Sigmoid()( nn.CAddTable(){w_t, w_s} )
    end

Shouldn't be a _model_ that holds the weights be returned and optimization algorithm applied? (lstm is only a _function_ in this case.) Or I just misunderstood here. Thanks.

**Comment by karpathy** [comment]
I think you probably misunderstand how Torch and neural nets work. I'd recommend a tutorial before diving into this code base. The functions you pointed to only execute a single time in the beginning and set up the computational graph, which is then traversed back and forth during training in the forward and backward calls. The params get initialized once only in the beginning.
